File: mobile/lib/models/product.dart
/// Mirrors the backend `ProductListItem` / `ProductDetail` schemas.
///
/// Money arrives as an exact decimal STRING and is kept as one. Parsing it into
/// a `double` would reintroduce the rounding error the backend's Decimal
/// columns exist to prevent. Format it with `core/money.dart`.
library;

enum StockStatus {
  inStock,
  lowStock,
  outOfStock;

  static StockStatus fromJson(String? value) => switch (value) {
        'IN_STOCK' => StockStatus.inStock,
        'LOW_STOCK' => StockStatus.lowStock,
        _ => StockStatus.outOfStock,
      };

  String get label => switch (this) {
        StockStatus.inStock => 'In stock',
        StockStatus.lowStock => 'Only a few left',
        StockStatus.outOfStock => 'Out of stock',
      };
}

class ProductImage {
  const ProductImage({required this.id, required this.imageUrl, this.altText});

  factory ProductImage.fromJson(Map<String, dynamic> json) => ProductImage(
        id: json['id'] as int,
        imageUrl: json['image_url'] as String,
        altText: json['alt_text'] as String?,
      );

  final int id;
  final String imageUrl;
  final String? altText;
}

class ProductCategory {
  const ProductCategory({required this.id, required this.name, required this.slug});

  factory ProductCategory.fromJson(Map<String, dynamic> json) => ProductCategory(
        id: json['id'] as int,
        name: json['name'] as String,
        slug: json['slug'] as String,
      );

  final int id;
  final String name;
  final String slug;
}

class Product {
  const Product({
    required this.id,
    required this.name,
    required this.slug,
    required this.sku,
    required this.category,
    required this.mrp,
    required this.sellingPrice,
    required this.discountPercentage,
    required this.discountAmount,
    required this.stockStatus,
    required this.inStock,
    this.primaryImageUrl,
    this.isFeatured = false,
    this.description,
    this.stockQuantity,
    this.images = const [],
  });

  factory Product.fromJson(Map<String, dynamic> json) => Product(
        id: json['id'] as int,
        name: json['name'] as String,
        slug: json['slug'] as String,
        sku: json['sku'] as String,
        category: ProductCategory.fromJson(json['category'] as Map<String, dynamic>),
        // Deliberately read as String, never num.
        mrp: json['mrp'] as String,
        sellingPrice: json['selling_price'] as String,
        discountPercentage: json['discount_percentage'] as String,
        discountAmount: json['discount_amount'] as String,
        stockStatus: StockStatus.fromJson(json['stock_status'] as String?),
        inStock: (json['in_stock'] as bool?) ?? false,
        primaryImageUrl: json['primary_image_url'] as String?,
        isFeatured: (json['is_featured'] as bool?) ?? false,
        description: json['description'] as String?,
        stockQuantity: json['stock_quantity'] as int?,
        images: (json['images'] as List<dynamic>?)
                ?.map((image) => ProductImage.fromJson(image as Map<String, dynamic>))
                .toList() ??
            const [],
      );

  final int id;
  final String name;
  final String slug;
  final String sku;
  final ProductCategory category;
  final String mrp;
  final String sellingPrice;
  final String discountPercentage;
  final String discountAmount;
  final StockStatus stockStatus;
  final bool inStock;
  final String? primaryImageUrl;
  final bool isFeatured;

  // Detail-only fields.
  final String? description;
  final int? stockQuantity;
  final List<ProductImage> images;

  /// Every image URL for the gallery, falling back to the card image.
  List<String> get galleryUrls {
    if (images.isNotEmpty) return images.map((image) => image.imageUrl).toList();
    final primary = primaryImageUrl;
    return primary == null ? const [] : [primary];
  }

  /// How many units a shopper may put in the basket at once.
  int get maxSelectableQuantity {
    final available = stockQuantity;
    if (available == null) return inStock ? 10 : 0;
    return available < 10 ? available : 10;
  }
}

File: backend/app/schemas/product.py
"""Product API schemas.

Money is exchanged as a Decimal, which Pydantic serialises to an exact JSON
string (e.g. "1250.00"). Clients must parse it into a decimal type - parsing
into a float reintroduces the rounding error the Decimal column exists to avoid.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.enums import StockStatus
from app.schemas.category import CategorySummary

# Matches the Numeric(10, 2) columns in the database.
Money = Annotated[Decimal, Field(max_digits=10, decimal_places=2, ge=Decimal("0"))]


def _clean_text(value: str) -> str:
    return " ".join(value.split())


# ---- Images -----------------------------------------------------------------
class ProductImageCreate(BaseModel):
    image_url: str = Field(min_length=1, max_length=500)
    alt_text: str | None = Field(default=None, max_length=200)
    display_order: int = Field(default=0, ge=0)
    is_primary: bool = False


class ProductImageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    image_url: str
    alt_text: str | None = None
    display_order: int
    is_primary: bool


# ---- Products ---------------------------------------------------------------
class ProductBase(BaseModel):
    name: str = Field(min_length=2, max_length=200)
    description: str | None = None
    category_id: int = Field(gt=0)
    mrp: Money
    selling_price: Money
    stock_quantity: int = Field(default=0, ge=0)
    low_stock_threshold: int = Field(default=10, ge=0)
    is_active: bool = True
    is_featured: bool = False

    @field_validator("name")
    @classmethod
    def _tidy_name(cls, value: str) -> str:
        cleaned = _clean_text(value)
        if len(cleaned) < 2:
            raise ValueError("Product name must be at least 2 characters long.")
        return cleaned

    @model_validator(mode="after")
    def _selling_price_within_mrp(self) -> ProductBase:
        # Mirrors the database CHECK constraint so the client gets a readable
        # 422 instead of a 500 from a constraint violation.
        if self.selling_price > self.mrp:
            raise ValueError("Selling price cannot be greater than the MRP.")
        return self


class ProductCreate(ProductBase):
    """Admin product creation.

    `sku` is optional: when omitted the service generates a unique one. The
    slug is always derived server-side.
    """

    sku: str | None = Field(default=None, min_length=2, max_length=64)
    images: list[ProductImageCreate] = Field(default_factory=list)

    @field_validator("sku")
    @classmethod
    def _normalise_sku(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = value.strip().upper()
        return cleaned or None


class ProductUpdate(BaseModel):
    """Partial update. Every field is optional."""

    name: str | None = Field(default=None, min_length=2, max_length=200)
    description: str | None = None
    category_id: int | None = Field(default=None, gt=0)
    sku: str | None = Field(default=None, min_length=2, max_length=64)
    mrp: Money | None = None
    selling_price: Money | None = None
    stock_quantity: int | None = Field(default=None, ge=0)
    low_stock_threshold: int | None = Field(default=None, ge=0)
    is_active: bool | None = None
    is_featured: bool | None = None

    @field_validator("name")
    @classmethod
    def _tidy_name(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = _clean_text(value)
        if len(cleaned) < 2:
            raise ValueError("Product name must be at least 2 characters long.")
        return cleaned

    @field_validator("sku")
    @classmethod
    def _normalise_sku(cls, value: str | None) -> str | None:
        return value.strip().upper() if value else None

    @model_validator(mode="after")
    def _selling_price_within_mrp(self) -> ProductUpdate:
        # Only checkable here when both arrive together; the service re-checks
        # against the stored value when only one is supplied.
        both_supplied = self.mrp is not None and self.selling_price is not None
        if both_supplied and self.selling_price > self.mrp:
            raise ValueError("Selling price cannot be greater than the MRP.")
        return self


class StockAdjustment(BaseModel):
    """Set or change stock. Exactly one of `set_to` / `delta` must be given."""

    set_to: int | None = Field(default=None, ge=0, description="Absolute new stock level.")
    delta: int | None = Field(default=None, description="Relative change, may be negative.")
    reason: str | None = Field(default=None, max_length=200)

    @model_validator(mode="after")
    def _exactly_one(self) -> StockAdjustment:
        if (self.set_to is None) == (self.delta is None):
            raise ValueError("Provide exactly one of 'set_to' or 'delta'.")
        return self


class ProductListItem(BaseModel):
    """Catalogue card projection - what a product grid needs, nothing more."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    slug: str
    sku: str
    category: CategorySummary
    mrp: Decimal
    selling_price: Decimal
    discount_percentage: Decimal
    discount_amount: Decimal
    stock_status: StockStatus
    in_stock: bool
    primary_image_url: str | None = None
    is_featured: bool
    is_active: bool


class ProductDetail(ProductListItem):
    """Full product view, including images and the exact stock level.

    `stock_quantity` is exposed so the mobile quantity selector can cap itself
    at what is actually available.
    """

    description: str | None = None
    stock_quantity: int
    images: list[ProductImageRead] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class ProductAdminDetail(ProductDetail):
    """Adds the operational fields only staff need to see."""

    low_stock_threshold: int
    sold_quantity: int
